test/test_unification.cpp

```diff
@@ -1,30 +1,223 @@
-    fol::Constant c{"c"};
-    // Test whether an empty substitution returns failure
-    std::optional<Substitution> theta_failure = std::nullopt; 
-    BOOST_TEST(!(unify(c, c, theta_failure)).has_value());
-
-    // Testing whether a non-empty substitution works
-    std::optional<Substitution> theta = Substitution();
-    theta.value()["c"] = fol::Constant{"c_subst"};
-    auto subst = unify(c, c, theta);
-
-    BOOST_TEST(subst.has_value());
-    BOOST_TEST(subst.value()["c"].name == "c_subst");
-
-    // Check unification with variables
-    fol::Variable v{"v"};
-    auto subst_2 = unify(v, c);
-    //BOOST_TEST(subst_2.value()["v"].name == "c");
+
+    /* --------- List of Test cases for unification -----------
+    Variables: v1, v2
+    Constants: C2, C3
+    Predicates: T(), F1(), E1() 
+
+    1. T(v1) and T(C2)
+    2. T(v1, C3) and T(C2, C3)
+    3. T(v1, v2) and T(C2, C3)
+    4. T(C2, C3) and T(v1, C3)
+    5. T(F1(v1)) and T(F1(C2))
+    6. T(v1) and T(F1(C2)) -------> Here is a reason to make the sub_list a string -> term mapping since the arguements are lost here.
+    7. T(E1(F1(v1)), C3) and T(E1(F1(C2)), v2)
+    8. Parsing unfication tests predicate(con_1, var_1) and predicate(con_1, con_2)
+    9. Here is a fail test case where unification is not possible: 
+        9.1 Different number of argument of literal 
+        9.2 No variable is given
+        9.3 different nested predicates
+    */
+
+    // BOOST_TEST(true);
+    // setup some definitions
+    typedef unordered_map<string, Term> sub_list_type;
+    Literal<Term> L1, L2;
+    Term T1, T2, T3, T4, T5, T6, T7, T8;
+    Constant C1, C2, C3;
+    Variable v1, v2;
+    Function F1, F2, F3, F4;
+    Predicate P1, P2;
+    sub_list test;
+    sub_list answer;
+
+    // sample predicates for testing unification
+
+    // T(V1) unified with T(C2)
+    // first literal declaration, T(v1)
+    v1.name = "v1";
+    T1 = v1;
+    P1 = "P1";
+    L1.predicate = P1;
+    L1.args.push_back(T1);
+    // second literal declaration, T(C2)
+    C2.name = "C2";
+    T2 = C2;
+    P2 = "P2";
+    L2.predicate = P2;
+    L2.args.push_back(T2);
+    // correct answer declaration, v1->C2
+    get<sub_list_type>(answer)["v1"] = C2;
+    // run test for single argument constant unification
+    test = unification(L1, L2, test);
+    BOOST_CHECK_EQUAL(get<Constant>(get<sub_list_type>(test)["v1"]).name, get<Constant>(get<sub_list_type>(answer)["v1"]).name);
+
+    // lets now up the number of arguments, first an extra constant that is already unified
+    C3.name = "C3";
+    T3 = C3;
+    L1.args.push_back(T3);
+    L2.args.push_back(T3);
+    // clear previous substitution list, answer is the same, no need to clear it 
+    get<sub_list_type>(test).clear();
+    // test unification of T(v1, C3) and T(C2, C3), v1->C2
+    test = unification(L1, L2, test);
+    BOOST_CHECK_EQUAL(get<Constant>(get<sub_list_type>(test)["v1"]).name, get<Constant>(get<sub_list_type>(answer)["v1"]).name);
```

Comment:
Is there a particular reason you're using `BOOST_CHECK_EQUAL` instead of `BOOST_TEST` in the test module?

Reply:
Originally I was trying to compare the objects in different ways and it was little more forgiving, but I'm just comparing strings now so I'll change it to BOOST_TEST.